### ISSUE TITLE
Predicate returns wrong data

### ISSUE THREAD

**Issue by carlo-activia** [hazelcast/hazelcast#9798]
When using EntryProcessor and index on a field the EntryProcessor uses I get wrong data when executing a Predicate. Here is the code:

  @Test
  public void issue() throws Exception {
    Config config = new Config();
    MapConfig testMapConfig = config.getMapConfig("testMap");
    testMapConfig.setInMemoryFormat(InMemoryFormat.OBJECT);
    testMapConfig.getMapIndexConfigs().add(new MapIndexConfig("lastValue", true));
    HazelcastInstance instance = Hazelcast.newHazelcastInstance(config);

    try {
      IMap<Object, Object> testMap = instance.getMap("testMap");

      testMap.set(1L, new MyData(10));

      testMap.executeOnKey(1L, new MyProcessor(15));

      Predicate betweenPredicate = Predicates.between("lastValue", 0, 11);
      Collection<Object> values = testMap.values(betweenPredicate);
      Assert.assertEquals(0, values.size());
    } finally {
      instance.shutdown();
    }
  }

Here is the EntryProcessor:

  static class MyProcessor extends AbstractEntryProcessor<Long, MyData> {
    private long newValue;

    public MyProcessor(long newValue) {
      this.newValue = newValue;
    }

    @Override
    public Object process(Map.Entry<Long, MyData> entry) {
      MyData data = entry.getValue();
      data.setLastValue(newValue);
      entry.setValue(data);
      return null;
    }
  }

I expect the assertion to work, but it fails I get 1 record. If I change OBJECT to BINARY or if I comment the line "testMapConfig.getMapIndexConfigs().add(new MapIndexConfig("lastValue", true));" then it works.

Any ideas? I am using Hazelcast 3.7.5.

Thanks in advance.

**Comment by tombujok** [comment]
Thx for reporting this.
Everything will work if you change your EntryProcessor to create a new entry with the new value.
```
static class MyProcessor extends AbstractEntryProcessor<Long, MyData> {
        private long newValue;

        public MyProcessor(long newValue) {
            this.newValue = newValue;
        }

        @Override
        public Object process(Map.Entry<Long, MyData> entry) {
            MyData newData = new MyData(newValue);
            entry.setValue(newData);
            return newValue;
        }
    }
```

In order to modify the index we need the old `entry.getValue()`. If you modify it the index won't find the old value in the Object since it has been modified.
With BINARY the object is deserialised so the oldValue is preserved.

I need to investigate tomorrow if this is intended behaviour.

**Comment by carlo-activia** [comment]
@tombujok : 
Thanks for the clarification and the suggested code. I tried it and my test passes now :)
BTW, let me know whether it is a bug or an intended behavior.

**Comment by tombujok** [comment]
It was a bug. Fixed via #9818
